SpatialGDK/Source/SpatialGDK/Private/Utils/EntityPool.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/EntityPool.h"
+#include "SpatialGDKSettings.h"
+#include "TimerManager.h"
+
+#include "Interop/SpatialReceiver.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialEntityPool);
+
+using namespace improbable;
+
+void UEntityPool::Init(USpatialNetDriver* InNetDriver, FTimerManager* InTimerManager)
+{
+	NetDriver = InNetDriver;
+	Receiver = InNetDriver->Receiver;
+	TimerManager = InTimerManager;
+
+	if (NetDriver->IsServer())
+	{
+		ReserveEntityIDs(GetDefault<USpatialGDKSettings>()->EntityPoolInitialReservationCount);
+	}
+}
+
+void UEntityPool::ReserveEntityIDs(int32 EntitiesToReserve)
+{
+	UE_LOG(LogSpatialEntityPool, Log, TEXT("Sending bulk entity ID Reservation Request"));
+
+	checkfSlow(!bIsAwaitingResponse, TEXT("Trying to reserve Entity IDs while another reserve request is in flight"));
+
+	// Set up reserve IDs delegate
+	ReserveEntityIDsDelegate CacheEntityIDsDelegate;
+	CacheEntityIDsDelegate.BindLambda([EntitiesToReserve, this](Worker_ReserveEntityIdsResponseOp& Op)
+	{
+		if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+		{
+			// UNR-630 - Temporary hack to avoid failure to reserve entities due to timeout on large maps
+			if (Op.status_code == WORKER_STATUS_CODE_TIMEOUT)
+			{
+				UE_LOG(LogSpatialEntityPool, Warning, TEXT("Failed to reserve entity IDs Reason: %s. Retrying..."), UTF8_TO_TCHAR(Op.message));
+				ReserveEntityIDs(EntitiesToReserve);
+			}
+			else
+			{
+				UE_LOG(LogSpatialEntityPool, Error, TEXT("Failed to reserve entity IDs Reason: %s."), UTF8_TO_TCHAR(Op.message));
+			}
+
+			return;
+		}
+
+		// Ensure we have the same number of reserved IDs as we have entities to spawn
+		check(EntitiesToReserve == Op.number_of_entity_ids);
+
+		// Clean up any expired Entity ranges
+		ReservedEntityIDRanges = ReservedEntityIDRanges.FilterByPredicate([](const EntityRange& Element)
+		{
+			return !Element.bExpired;
+		});
+
+		EntityRange NewEntityRange = {};
+		NewEntityRange.CurrentEntityId = Op.first_entity_id;
+		NewEntityRange.LastEntityId = Op.first_entity_id + (Op.number_of_entity_ids - 1);
+		NewEntityRange.EntityRangeId = NextEntityRangeId++;
+
+		UE_LOG(LogSpatialEntityPool, Log, TEXT("Reserved %d entities, caching in pool, Entity IDs: (%d, %d) Range ID: %d"), Op.number_of_entity_ids, Op.first_entity_id, NewEntityRange.LastEntityId, NewEntityRange.EntityRangeId);
+
+		ReservedEntityIDRanges.Add(NewEntityRange);
```

Comment:
Any reason to not use a Map for this structure, with the EntityRangeId as the key?

Reply:
There's only ever 2-3 of these.